timemachine/cpp/src/nonbonded_pairs.cu

```diff
@@ -0,0 +1,197 @@
+void NonbondedPairs<RealType, Negated, Interpolated>::execute_device(
+    const int N,
+    const int P,
+    const double *d_x,
+    const double *d_p,
+    const double *d_box,
+    const double lambda,
+    unsigned long long *d_du_dx,
+    unsigned long long *d_du_dp,
+    unsigned long long *d_du_dl,
+    unsigned long long *d_u,
+    cudaStream_t stream) {
+
+    const int tpb = 32;
+
+    int num_blocks = ceil_divide(N, tpb);
+    dim3 dimGrid(num_blocks, 3, 1);
+
+    CUresult result = compute_w_coords_instance_.configure(dimGrid, tpb, 0, stream)
+                          .launch(N, lambda, cutoff_, d_lambda_plane_idxs_, d_lambda_offset_idxs_, d_w_, d_dw_dl_);
+    if (result != 0) {
+        throw std::runtime_error("Driver call to k_compute_w_coords");
+    }
+
+    if (Interpolated) {
+        CUresult result = compute_permute_interpolated_.configure(dimGrid, tpb, 0, stream)
+                              .launch(lambda, N, d_perm_, d_p, d_p_interp_, d_dp_dl_);
+        if (result != 0) {
+            throw std::runtime_error("Driver call to k_permute_interpolated failed");
+        }
+    } else {
+        gpuErrchk(cudaMemsetAsync(d_dp_dl_, 0, N * 3 * sizeof(*d_dp_dl_), stream))
+    }
+
+    if (d_du_dp) {
+        gpuErrchk(cudaMemsetAsync(d_du_dp_buffer_, 0, N * 3 * sizeof(*d_du_dp_buffer_), stream))
+    }
+
+    int num_blocks_pairs = ceil_divide(M_, tpb);
+
+    k_nonbonded_pairs<RealType, Negated><<<num_blocks_pairs, tpb, 0, stream>>>(
+        M_,
+        d_x,
+        Interpolated ? d_p_interp_ : d_p,
+        d_box,
+        d_dp_dl_,
+        d_w_,
+        d_dw_dl_,
+        d_pair_idxs_,
+        d_scales_,
+        beta_,
+        cutoff_,
+        d_du_dx,
+        d_du_dp_buffer_,
+        d_du_dl,
+        d_u);
+
+    gpuErrchk(cudaPeekAtLastError());
+
+    if (d_du_dp) {
+        if (Interpolated) {
+            CUresult result = compute_add_du_dp_interpolated_.configure(dimGrid, tpb, 0, stream)
+                                  .launch(lambda, N, d_du_dp_buffer_, d_du_dp);
+            if (result != 0) {
+                throw std::runtime_error("Driver call to k_add_du_dp_interpolated failed");
+            }
+        } else {
+            k_add_ull_to_ull<<<dimGrid, tpb, 0, stream>>>(N, d_du_dp_buffer_, d_du_dp);
+        }
+        gpuErrchk(cudaPeekAtLastError());
+    }
+}
+
+// TODO: this implementation is duplicated from NonbondedDense. Worth adding NonbondedBase?
+template <typename RealType, bool Negated, bool Interpolated>
+void NonbondedPairs<RealType, Negated, Interpolated>::du_dp_fixed_to_float(
+    const int N, const int P, const unsigned long long *du_dp, double *du_dp_float) {
+
+    // In the interpolated case we have derivatives for the initial and final parameters
+    const int num_tuples = Interpolated ? N * 2 : N;
+
+    for (int i = 0; i < num_tuples; i++) {
+        const int idx_charge = i * 3 + 0, idx_sig = i * 3 + 1, idx_eps = i * 3 + 2;
```

Comment:
nit: Having this on three lines would be easier to read

Reply:
I like it, changed in 935428b